src/lib/OS/PageTable.cc

```diff
@@ -0,0 +1,129 @@
+#include "simeng/OS/PageTable.hh"
+
+#include <cmath>
+#include <iostream>
+#include <vector>
+
+#include "simeng/util/Math.hh"
+
+namespace simeng {
+namespace OS {
+
+PageTable::PageTable(){};
+
+PageTable::~PageTable(){};
+
+bool PageTable::isMapped(uint64_t vaddr) { return find(vaddr) != table_.end(); }
+
+uint64_t PageTable::generateOffsetMask(uint64_t pageSize) {
+  if (!isPow2(page_size)) {
+    std::cerr
+        << "[SimEng:PageTable] Page size should be aligned to a power of 2."
+        << std::endl;
+    std::exit(1);
+  }
+  uint64_t lval = std::log2(page_size);
+  uint64_t mask = 0;
+  mask = ~mask;
+  mask ^= (mask << lval);
+  return mask;
+}
+
+PageTable::TableItr PageTable::find(uint64_t vaddr) {
+  uint64_t lowestPageStart = downAlign(vaddr, page_size);
+  TableItr itr = table_.find(lowestPageStart);
+  return itr;
+};
+
+void PageTable::ignoreAddrRange(uint64_t startAddr, uint64_t endAddr) {
+  ignored_ = std::pair<uint64_t, uint64_t>(startAddr, endAddr);
+  return;
+};
+
+void PageTable::allocatePTEntry(uint64_t alignedVAddr, uint64_t phyAddr) {
+  table_.insert(std::pair<uint64_t, uint64_t>(alignedVAddr, phyAddr));
+};
+
+void PageTable::deletePTEntry(PageTable::TableItr itr) { table_.erase(itr); };
+
+uint64_t PageTable::calculateOffset(uint64_t vaddr) {
+  // A mask of 0b0111111111111 to get the lower 12 bits of the vaddr.
```

Comment:
SimEng currently only supports 4096 bytes as page_size, it will be confusing if a generic comment is added. I have added a TODO on this comment as a reminder to change this when the change to  user-specified page size is made.